client/asset/eth/multirpc.go

```diff
@@ -784,11 +781,26 @@ func errorFilter(err error, matches ...interface{}) bool {
 }
 
 // withOne runs the provider function against the providers in order until one
-// succeeds or all have failed.
-func (m *multiRPCClient) withOne(providers []*provider, f func(*provider) error, acceptabilityFilters ...acceptabilityFilter) (superError error) {
+// succeeds or all have failed. The context used to run functions has a time
+// limit equal to defaultRequestTimeout for all requests to return. If
+// operations are expected to run longer than that the calling function should
+// not use the altered context.
+func (m *multiRPCClient) withOne(ctx context.Context, providers []*provider, f func(context.Context, *provider) error, acceptabilityFilters ...acceptabilityFilter) (superError error) {
 	readyProviders := make([]*provider, 0, len(providers))
 	for _, p := range providers {
 		if !p.failed() {
+			ctx, cancel := context.WithTimeout(ctx, defaultRequestTimeout)
+			// Fetching the best header will check that either the
+			// provider's cached header is not too old or that a
+			// newly fetched header is not too old. If it is too
+			// old that indicates the provider is not in sync and
+			// should not be used.
+			_, err := p.bestHeader(ctx, m.log)
+			cancel()
+			if err != nil {
+				m.log.Warnf("Problem getting best header: %s.", err)
+				continue
+			}
```

Comment:
`bestHeader` is called on a random provider from the list every second. I think we can rely on that to keep `failed` up to date.

Reply:
You mean in `checkForNewBlocks`?

I guess it depends on how many providers... it would not be guaranteed unless one provider.

Reply:
Can we do this check down in the `for _, p := range readyProviders` loop right before we actually use the provider, so that if the first one passes we're not wasting calls on the rest?

Reply:
Moved to the loop.